### ISSUE TITLE
[RM Chxx] Service function chaining

### ISSUE THREAD

**Issue by walterkozlowski** [cntt-n/CNTT#2089]
As per the gap description

**Comment by sureshkrishnan** [comment]
Over the past few years there has been a significant move towards decomposing network functions into smaller sub-functions that can be independently scaled and potentially reused across multiple network functions. A service chain allows composition of network functions by passing selected packets through an ordered list of services.

In order to support this capability in a sustainable manner, there is a need to have the capability to model service chains as a high level abstraction. This is essential to ensure that the underlying connection setup, and (re-)direction of traffic flows can be performed in an automated manner. At a very high level a service chain can be considered a directed acyclic graph with the composing network functions being the vertices. Building on top of this, a service chain can be modeled by defining two parameters

- An ordered list of service functions that need to be traversed for a service chain
- A set of packet/flow classifiers that determine what packets will enter a given service chain

These capabilities need to be provided for both virtualized and containerized (cloud-native) network functions as there will be a need to support both of them for the foreseeable future. Since virtualized network functions have existed for a while there is existing support for service chaining in virtualized environments in orchestration platforms like Openstack. Container orchestration platforms such as Kubernetes don't support service chaining and may require development of new primitives in order to support advanced networking functions.

There is also a need to provide specialized tools aid troubleshooting of individual services and the communication between them in order to investigate issues in the performance of composed network functions. Minimally, there is a need to provide packet level and byte level counters and statistics as the packets pass through the service chain in order to ascertain any issues with forwarding and performance. Additionally, there is a need for mechanisms to trace the paths of selected subsets of traffic as they flow through the service chain.

**Comment by pgoyal01** [comment]
@sureshkrishnan Nice start.

Some comments:
1. " a service chain can be considered a directed acyclic graph" and "A service chain allows composition of network functions by passing selected packets through an ordered list of services" -- the acyclic graph is a better abstraction in that it allows multiple paths. The OpenStack plugin is very restrictive a single defined flow instead of a single ingress and a single egress but with multiple (maybe different) cost paths (either service functions or the instances (on different VMs/containers)) in between.
2. Should the abstraction consider a workflow manager that executes an SFC acyclic graph?

**Comment by sureshkrishnan** [comment]
@pgoyal01 , really good points Pankaj. How about I change the text as follows

OLD:
An ordered list of service functions that need to be traversed for a service chain

NEW:
An acyclic graph defining the service functions that need to be traversed for the service chain. This allows for multiple paths for a packet to traverse the service chain.

Proposal for workflow manager text to be added at the end of paragraph 3 or beginning of paragraph 4 (last paragraph). Please propose improvements/alternate text if needed

"It is expected that reference architectures will provide a service chain workflow manager that would accept the service function acyclic graph and be able to identify/create the necessary service functions and the networking between them in order to instantiate such a chain."

**Comment by sureshkrishnan** [comment]
With the proposed changes, the text would read as follows:

Over the past few years there has been a significant move towards decomposing network functions into smaller sub-functions that can be independently scaled and potentially reused across multiple network functions. A service chain allows composition of network functions by passing selected packets through an ordered list of services.

In order to support this capability in a sustainable manner, there is a need to have the capability to model service chains as a high level abstraction. This is essential to ensure that the underlying connection setup, and (re-)direction of traffic flows can be performed in an automated manner. At a very high level a service chain can be considered a directed acyclic graph with the composing network functions being the vertices. Building on top of this, a service chain can be modeled by defining two parameters

- An acyclic graph defining the service functions that need to be traversed for the service chain. This allows for multiple paths for a packet to traverse the service chain.

- A set of packet/flow classifiers that determine what packets will enter a given service chain

These capabilities need to be provided for both virtualized and containerized (cloud-native) network functions as there will be a need to support both of them for the foreseeable future. Since virtualized network functions have existed for a while there is existing support for service chaining in virtualized environments in orchestration platforms like Openstack. Container orchestration platforms such as Kubernetes don't support service chaining and may require development of new primitives in order to support advanced networking functions.

It is expected that reference architectures will provide a service chain workflow manager that would accept the service function acyclic graph and be able to identify/create the necessary service functions and the networking between them in order to instantiate such a chain.

There is also a need to provide specialized tools aid troubleshooting of individual services and the communication between them in order to investigate issues in the performance of composed network functions. Minimally, there is a need to provide packet level and byte level counters and statistics as the packets pass through the service chain in order to ascertain any issues with forwarding and performance. Additionally, there is a need for mechanisms to trace the paths of selected subsets of traffic as they flow through the service chain.

**Comment by pgoyal01** [comment]
@sureshkrishnan Looks good.
The OpenStack SFC is not a full solution.

BTW CNTT uses British English.

**Comment by walterkozlowski** [comment]
@sureshkrishnan, I like it.  I think it is ready for a PR.  In my view, this could go to RM 2.3 Use Cases.  Your thoughts?

**Comment by kumarsumangit** [comment]
@sureshkrishnan ,  i have two queries - 1. we are missing that SFC enabled domain definition where SFC will be applicable and it need to be mentioned. please ignore if it is not required here. 
2. "Building on top of this, a service chain can be modeled by defining two parameters" -> According to the SFC RFC, it may required Service Function forwarder to enable packet/frames/flow movement to SF as these are critical component for SFC along with SF.  Do we need to mention this as another parameters or it will be part of acyclic graph of SF(s) and SFF(s)?

**Comment by sureshkrishnan** [comment]
@pgoyal01 : Ack. I will mention that the Openstack solution is partial and will change the spelling to British English :-).

@walterkozlowski :  sounds good to me. Not sure if you want to discuss the location in the meeting today but in either case I will send a PR for the right location by end of today.

@kumarsumangit  : I think there might be a slight misunderstanding here. We can discuss if we need new text. I used the term "Service function chaining" to describe the concept itself and not the specific instance of it that is the IETF SFC effort. That is one of the key reasons I did not use the acronym. If you think further clarifications are needed, we can discuss some text.
